### PR TITLE
Fix for universal autolog configs overriding integration-specific configs

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

A bug in the code for the `setup_autolog` post import hook for libraries meant that the `disable` flag was only respected when the order of:
- importing the library
- `mlflow.autolog()`
- `mlflow.integration.autolog(disable=True)`

is in a specific order. In particular, the `setup_autolog` function called `mlflow.integration.autolog` as a post-import hook, but with the _default_ values for its params, rather than the ones that may have already been set as stored in `AUTOLOGGING_INTEGRATIONS`.
Fix by only calling `integration.autolog()` with `mlflow.autolog`'s configs, if there is no pre-existing integration-specific configs (ie. `mlflow.integration.autolog` has not been called.

Summary of manually verified fixes:
1. mlflow.autolog, then library autolog disable, then import library
2. mlflow.autolog, then import library, then library autolog disable
3. library autolog disable, then mlflow.autolog, then import library
4. library autolog disable, then import library, then mlflow.autolog
5. import library, then mlflow.autolog, then library autolog disable
6. import library, then library autolog disable, then mlflow.autolog
## How is this patch tested?

(Details)

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fix a bug causing configurations passed to `mlflow.autolog()` to override configurations passed to framework-specific autologging functions  (e.g., `mlflow.sklearn.autolog()`). Framework-specific autologging configurations will now take precedence over all universal `mlflow.autolog()` configurations.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -1339,8 +1345,27 @@ def get_autologging_params(autolog_fn):
     def setup_autologging(module):
         try:
             autolog_fn = LIBRARY_TO_AUTOLOG_FN[module.__name__]
+
+            # Only call integration's autolog function with `mlflow.autolog` configs
+            # if the integration's autolog function has not already been called.
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py:1350]:
```suggestion
            # if the integration's autolog function has not already been called directly by the user.
```

File: mlflow/tracking/fluent.py [hunk 2]
```diff
@@ -1296,6 +1300,26 @@ def print_auto_logged_info(r):
...
+    but
+
+    .. code-block:: python
+
+        mlflow.autolog(log_models=False, exclusive=True)
+        import sklearn
+        mlflow.sklearn.autolog(log_models=True)
+
+    would enable autologging for `sklearn` with `log_models=True` and `exclusive=False`,
+    the latter resulting from the default value for `exclusive` in `mlflow.sklearn.autolog`.
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py:1322]: We should move this up above the params and code snippet  (after line 1234).

File: mlflow/tracking/fluent.py [hunk 3]
```diff
@@ -1296,6 +1300,26 @@ def print_auto_logged_info(r):
...
+    would enable autologging for `sklearn` with `log_models=False` and `exclusive=True`,
+    but
+
+    .. code-block:: python
+
+        mlflow.autolog(log_models=False, exclusive=True)
+        import sklearn
+        mlflow.sklearn.autolog(log_models=True)
+
+    would enable autologging for `sklearn` with `log_models=True` and `exclusive=False`,
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py:1321]: can we clarify that all other frameworks will use the configuration `log_models=False` and `exclusive=True`?

File: mlflow/tracking/fluent.py (hunk 2 above)
Comment (apurva-koti) [reply mlflow/tracking/fluent.py:1322]: Done!

File: mlflow/tracking/fluent.py (hunk 3 above)
Comment (apurva-koti) [reply mlflow/tracking/fluent.py:1321]: Yup!

**Review by dbczumar — COMMENTED** [review]
@apurva-koti Your commits to this PR LG to me! Feel free to approve and merge the PR once you're satisfied with it!

File: mlflow/utils/autologging_utils.py [hunk 4]
```diff
@@ -349,6 +349,7 @@ def autolog(*args, **kwargs):
             return _autolog(*args, **kwargs)
 
         wrapped_autolog = _update_wrapper_extended(autolog, _autolog)
+        wrapped_autolog.integration_name = name
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:352]: This sets a function attribute (https://www.python.org/dev/peps/pep-0232/) allowing us to fetch the autologging integration name from the framework-specific autologging function being examined in the `setup_autologging` webhook. I'll add an inline comment.

**Review by apurva-koti — APPROVED** [review]
good stuff @dbczumar

**Pull request merged**